src/Plugin/Commerce/PaymentGateway/HostedPaymentFields.php

```diff
@@ -771,4 +801,104 @@ protected function prepareSubscriptionChargeData(PaymentInterface $payment) {
+    $form['statement_descriptors']['card_types'] = [
+      '#markup' => '<p>' . $this->t(
+        'You may specify statement descriptors for specific card types below. If
+        no descriptor or support phone number is defined for a card type, the
+        global descriptor or phone number defined here (or the ones defined in
+        the BlueSnap account if no globals are defined here) will be used.'
+      ) . '</p>',
+    ];
+
+    $card_types = $this->getPluginDefinition()['credit_card_types'];
+    foreach ($card_types as $card_type) {
+      $form['statement_descriptors']['card_types'][$card_type] = [
+        '#type' => 'fieldset',
+        '#title' => $card_type,
+        'descriptor' => [
+          '#type' => 'textfield',
+          '#title' => $this->t('Soft descriptor'),
+          '#default_value' => $config[$card_type]['descriptor'],
+        ],
+        'phone_number' => [
+          '#type' => 'textfield',
+          '#title' => $this->t('Support phone number'),
+          '#default_value' => $config[$card_type]['phone_number'],
+        ],
+      ];
+    }
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function validateDescriptorFormElement(
+    array &$form,
+    FormStateInterface $form_state
+  ) {
+    parent::validateDescriptorFormElement($form, $form_state);
+
+    // Validate statement descriptors per card type.
+    $values = $form_state->getValue($form['#parents']);
+    $descriptors = &$values['statement_descriptors']['card_types'];
+
+    $card_types = $this->getPluginDefinition()['credit_card_types'];
+    foreach ($card_types as $card_type) {
+      $result = $this->validateDescriptor($descriptors[$card_type]['descriptor']);
+      if (!$result['valid']) {
+        $form_state->setError(
+          $form['statement_descriptors']['card_types'][$card_type]['descriptor'],
+          $result['error']
+        );
+      }
+
+      $result = $this->validateDescriptorPhoneNumber($descriptors[$card_type]['phone_number']);
+      if (!$result['valid']) {
+        $form_state->setError(
+          $form['statement_descriptors']['card_types'][$card_type]['phone_number'],
+          $result['error']
+        );
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function prepareDescriptorData(PaymentInterface $payment) {
+    $data = [];
+    $descriptors = &$this->configuration['statement_descriptors']['card_types'];
+    $card_type = $payment->getPaymentMethod()->get('card_type')->value;
+
+    // If there is a statement descriptor configured for the payment's card type
+    // use that.
+    if ($descriptors[$card_type]['descriptor']) {
+      $data['softDescriptor'] = $this->replaceDescriptorDataTokens(
+        $descriptors[$card_type]['descriptor'],
+        $payment
+      );
+    }
+    if ($descriptors[$card_type]['phone_number']) {
+      $data['descriptorPhoneNumber'] = $this->replaceDescriptorDataTokens(
```

Comment:
Do we need token replacement for the phone? Should this just be a 'telephone' field?

Reply:
I decided to allow tokens in the phone field as well so that you can do `[commerce_store:field_phone]`, it can be useful.